File: hard/distinct_subsequences/distinct.sub.cpp
// Reuben Thorpe (2016), CodeEval [Distinct Subsequence v1.1]
#include <iostream>
#include <fstream>
#include <vector>


using namespace std;
int gResult = 0;


vector<int> findIn(string inString, char match) {
  // Return a list of all positions char occures in string
  vector<int> indicies;
  for (int i = 0; i < inString.length(); i++) {
    if (inString[i] == match) {
      indicies.push_back(i);
    }
  }
  return(indicies);
}


void countSub(vector<vector<int>> freqArray, int depth, int cut) {
    // Counts the number of sub-sequences in the position matrix.
    // This function used global variable "gResult"
  if (depth == 0) {
    for (int i: freqArray[depth]) {
      countSub(freqArray, depth+1, i);
    }
  }
  else if (depth == freqArray.size()-1) {
    for (int i: freqArray[depth]) {
      if (i > cut) {
        gResult++;
      }
    }
  }
  else {
    for (int i: freqArray[depth]) {
      if (i > cut) {
        countSub(freqArray, depth+1, i);
      }
    }
  }
}


int main(int argc, char* argv[]) {
  // Generates position matrix, and calculates sub string frequency
  ifstream inFile(argv[1]);
  string seq, subSeq;

  while(getline(inFile, seq, ',') && getline(inFile, subSeq)) {
    vector<vector<int>> freqArray;

    for (char i: subSeq) {
      freqArray.push_back(findIn(seq, i));
    }

    countSub(freqArray, 0, 0);
    cout << gResult << "\n";
    gResult = 0;
  }
  return(0);
}

File: hard/distinct_subsequences/distinct.sub.py
# Reuben Thorpe (2016), CodeEval [Distinct Subsequence v1.0]
# NOTE : USING SETS MY BE FASTER BUT UNSTABLE, CONVERT TO CLASS
from sys import argv


def findIn(inString, char):
    # Returns all position indicies of char in string
    return([i for i, check in enumerate(inString) if check is char])


def distSub(seq, subSeq):
    """ Generates position matrix, and then counts the number of distinct
         sub-sequences with recursion. The print results
    """
    freqMatrix = []

    for char in subSeq:
        freqMatrix.append(findIn(seq, char))

    countSub(freqMatrix, 0, 0)
    print(gResult)


def countSub(freqMatrix, depth, cut):
    """ Counts the number of sub-sequences in the position matrix.
        This function used global variable "gResult" and "gLimit"
    """
    global gLimit
    if depth is 0:
        for i in freqMatrix[0]:
            countSub(freqMatrix, depth+1, i)

    elif depth is gLimit:
        global gResult
        for i in freqMatrix[depth]:
            if i > cut:
                gResult += 1

    else:
        for i in freqMatrix[depth]:
            if i > cut:
                countSub(freqMatrix, depth+1, i)


if __name__ == "__main__":
    gLimit = gResult = 0
    for line in open(argv[1], "r"):
        seq, subSeq = line.strip().split(',')
        gLimit = len(subSeq)-1
        distSub(seq, subSeq)
        gResult = 0;
